### PR TITLE
Cache backend API example

### PR DESCRIPTION
**Pull request by crazy-max**
Follow-up moby/buildkit#1974 docker/buildx#535

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

### CODE REVIEW

**Comment by crazy-max** [comment]
Only merge when buildx 0.6 and BuildKit 0.9 are released. In the meantime you can test this feature with the following workflow:

```yaml
name: ci

on:
  push:
    branches:
      - 'master'

jobs:
  docker:
    runs-on: ubuntu-latest
    steps:
      -
        name: Checkout
        uses: actions/checkout@v2
      -
        name: Set up Docker Buildx
        uses: docker/setup-buildx-action@v1
        with:
          version: "https://github.com/docker/buildx.git#67bd6f4dc82a9cd96f34133dab3f6f7af803bb14"
          driver-opts: |
            image=moby/buildkit:master
            network=host
          buildkitd-flags: --debug
      -
        name: Login to DockerHub
        uses: docker/login-action@v1 
        with:
          username: ${{ secrets.DOCKERHUB_USERNAME }}
          password: ${{ secrets.DOCKERHUB_TOKEN }}
      -
        name: Build and push
        uses: docker/build-push-action@v2
        with:
          context: .
          push: true
          tags: user/app:latest
          cache-from: type=gha
          cache-to: type=gha
```

**Comment by mmeendez8** [comment]
I think this can be merged now :tada:

**Comment by crazy-max** [comment]
@mmeendez8 [GitHub virtual environments](https://github.com/actions/virtual-environments) don't have buildx 0.6.0 ([0.5.1 atm](https://github.com/docker/build-push-action/runs/3096831142?check_suite_focus=true#step:5:4)) so will wait until then.

**Comment by mmeendez8** [comment]
Oh I see, any idea about when this might happen? We may create a new topic in a [discussions section](https://github.com/actions/virtual-environments/discussions)

**Comment by crazy-max** [comment]
> Oh I see, any idea about when this might happen? We may create a new topic in a discussions section

I think it's Microsoft internal. cc @cpuguy83

**Comment by cpuguy83** [comment]
Having some pipeline issues at the moment, but hoping to get this out today or tomorrow.... then it'll need to be picked up in the virtual-env image that GHA uses.

**Comment by crazy-max** [comment]
@cpuguy83 Thanks a bunch!

**Comment by cpuguy83** [comment]
0.6.0 is available now, but I'm not sure when the GHA virtual-env images will be updated. You should be able to manually update (`apt-get update && apt-get install -y moby-buildx`).

**Comment by ianschmitz** [comment]
This is awesome guys - great work!

We've been trying this out over the last couple days and have noticed that the first build for a PR typically doesn't seem to have any cached layers. However subsequent commits to the PR seem to have cached layers and are much faster. Is this expected?

For example my hope would have been that a PR merging into `main` would use the cache from `main` if one doesn't already exist for the PR.

**Comment by catthehacker** [comment]
> 0.6.0 is available now, but I'm not sure when the GHA virtual-env images will be updated.

Images are generated on a weekend and deployed throughout the week, so it might be next week when `0.6.0` will be widely available or in 2 weeks.

**Comment by robpc** [comment]
> This is awesome guys - great work!
> 
> We've been trying this out over the last couple days and have noticed that the first build for a PR typically doesn't seem to have any cached layers. However subsequent commits to the PR seem to have cached layers and are much faster. Is this expected?
> 
> For example my hope would have been that a PR merging into `main` would use the cache from `main` if one doesn't already exist for the PR.

@ianschmitz Are you using multiple restore keys? We use a system like this to fallback to a "any" image if the branch one isn't available.

```
      - name: Cache Docker layers
        uses: actions/cache@v2
        with:
          path: /tmp/.buildx-cache
          key: ${{ env.app_env }}-${{ github.job }}-${{ env.branch }}-${{ github.sha }}
          restore-keys: |
            ${{ env.app_env }}-${{ github.job }}-${{ env.branch }}-
            ${{ env.app_env }}-${{ github.job }}-
```

So in this example you could make the second one match your main branch key.

**Comment by ianschmitz** [comment]
@robpc negative. We're following the example laid out in this PR (i think it's correct?). It looks like this:

```yaml
      - name: Set up QEMU
        uses: docker/setup-qemu-action@v1

      - name: Set up Docker Buildx
        uses: docker/setup-buildx-action@v1
        with:
          version: v0.6.0-rc1
          driver-opts: image=moby/buildkit:v0.9.0-rc2
          buildkitd-flags: --debug

      - name: Login to ECR
        uses: docker/login-action@v1
        with:
          registry: ${{ env.DOCKER_REG }}
          username: ${{ secrets.AWS_ACCESS_KEY_ID }}
          password: ${{ secrets.AWS_SECRET_ACCESS_KEY }}

      - name: Build and push Docker image
        uses: docker/build-push-action@v2
        with:
          cache-from: type=gha
          cache-to: type=gha
          file: docker/Dockerfile
          platforms: linux/amd64, linux/arm64
          push: true
          tags: ${{ env.DOCKER_REG }}/foo:${{ github.sha }}
```

Where i was hoping that `cache-from: type=gha` would do roughly what i was hoping for above.

**Comment by robpc** [comment]
> @robpc negative. We're following the example laid out in this PR (i think it's correct?). It looks like this:
> 
> Where i was hoping that `cache-from: type=gha` would do roughly what i was thinking above.

@ianschmitz Sorry, I got my wires crossed a bit. The example I showed was from the workaround to using local in buildx and manually setting the gh cache. I am wondering if there is an equivalent way of specifying multiple keys in this buildx option. I realize now that is probably what your question was, so apologies for that. 

It sounds like `scope` is the way to specify that, but I didn't see in the [documentation](https://github.com/moby/buildkit/tree/master#github-actions-cache-experimental) if one could specify multiple and if that behaves like the `actions/cache@v2`.

**Comment by ianschmitz** [comment]
Sorry, I got my wires crossed a bit. The example I showed was from the workaround to using local in buildx and manually setting the gh cache. I am wondering if there is an equivalent way of specifying multiple keys in this buildx option. I realize now that is probably what your question was, so apologies for that.

Np! I guess that's what my question is? I was more hoping that the default behavior when using `type=gha` would be fairly intuitive where it pulls from target branch or default branch cache if it exists. I think that's what most folks would want/expect out of the box? From what i've experienced so far, the first job run in a new PR doesn't benefit from this caching.

**Comment by jojomatik** [comment]
> Effectively the pseudo logic i was thinking would be something like:
> 
> 1. Does a cache exist for this PR? If so, use it
> 2. Does a cache exist for the target branch? If so, use it
> 3. Does a cache exist for the default branch? If so, use it
> 
> Something along those lines 😃

@ianschmitz 
This is the logic that the "usual" (i.e. [actions/cache](https://github.com/actions/cache)) github actions cache follows [(link to documentation)](https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache):
> A workflow can access and restore a cache created in the current branch, the base branch (including base branches of forked repositories), or the default branch (usually main). For example, a cache created on the default branch would be accessible from any pull request. Also, if the branch feature-b has the base branch feature-a, a workflow triggered on feature-b would have access to caches created in the default branch (main), feature-a, and feature-b.
 
As far as I can tell from my workflows (e.g. https://github.com/jojomatik/blockcluster/pull/147 [(workflow)](https://github.com/jojomatik/blockcluster/runs/3106967125)) the same logic is true for this new cache backend. Not sure why it's not working for you.

**Comment by robpc** [comment]
> As far as I can tell from my workflows (e.g. [jojomatik/blockcluster#147](https://github.com/jojomatik/blockcluster/pull/147) [(workflow)](https://github.com/jojomatik/blockcluster/runs/3106967125)) the same logic is true for this new cache backend. Not sure why it's not working for you.

As far as I know the github cache action doesn't have a default. From the documentation for the cache action:

>The cache action will attempt to restore a cache based on the key you provide. When the action finds a cache, the action restores the cached files to the path you configure.
>
>If there is no exact match, the action creates a new cache entry if the job completes successfully. The new cache will use the key you provided and contains the files in the path directory.
>
>You can optionally provide a list of restore-keys to use when the key doesn't match an existing cache. 

The implication there is that the user is supplying both the `key` and the `restore-keys` which makes sense because depending on what you are doing they could be wildly different. Given that the action doesn't impose a default, I am wondering if this change is implementing one. I could see a reasonable default being sha -> branch -> repo level caching, but everyone's build setup is a little different so I am curious what the fallback logic is as implemented by this feature in buildx and if we have the power to change it. 

If this new feature in buildx is more or less mimicking this functionality, what is the `key` and `restore-keys` being used? From the [documentation](https://github.com/moby/buildkit/tree/master#github-actions-cache-experimental) on buildx, and a comment from @crazy-max  (I think) in an issue I can't find, it sounds like `scope` is maybe related to something buildx is doing with the keys.

**Comment by crazy-max** [comment]
> It sounds like `scope` is the way to specify that from the other comments, but I didn't see in the [documentation](https://github.com/moby/buildkit/tree/master#github-actions-cache-experimental) if one could specify multiple and if that behaves like

`scope` is documented in [`--export-cache`](https://github.com/moby/buildkit/tree/master#--export-cache-options) and [`--import-cache`](https://github.com/moby/buildkit/tree/master#--import-cache-options) sections. I think we will improve this doc and display all attributes for each cache type instead for better visibility.

**Comment by catthehacker** [comment]
> > 0.6.0 is available now, but I'm not sure when the GHA virtual-env images will be updated.
> 
> Images are generated on a weekend and deployed throughout the week, so it might be next week when `0.6.0` will be widely available or in 2 weeks.

Runners image update 20210726 includes Buildx `0.6.0`. PRs (section `Tools` > `Docker-Buildx`):
- [`ubuntu-18.04`](https://github.com/actions/virtual-environments/pull/3792)
  
  [![ubuntu-18.04-badge][ubuntu-18.04-badge]][ubuntu-18.04-status]
- [`ubuntu-20.04`](https://github.com/actions/virtual-environments/pull/3793)
  
  [![ubuntu-20.04-badge][ubuntu-20.04-badge]][ubuntu-20.04-status]

PRs should be merged once all GitHub runners are deployed with this version (when above badges will be at 100% with version `20210726`).
Mind you that during deployment phase you can land on a runner with update `20210718` or `20210726` and it's not possible to pin version.

[ubuntu-20.04]: https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md
[ubuntu-20.04-badge]: https://actionvirtualenvironmentsstatus.azurewebsites.net/api/status?imageName=ubuntu20&badge=1
[ubuntu-20.04-status]: https://actionvirtualenvironmentsstatus.azurewebsites.net/api/status?imageName=ubuntu20&redirect=1
[ubuntu-18.04]: https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu1804-README.md
[ubuntu-18.04-badge]: https://actionvirtualenvironmentsstatus.azurewebsites.net/api/status?imageName=ubuntu18&badge=1
[ubuntu-18.04-status]: https://actionvirtualenvironmentsstatus.azurewebsites.net/api/status?imageName=ubuntu18&redirect=1

**Comment by bekriebel** [comment]
I'm not sure if this is related to the `scope` discussion or not, but I'm finding that cache is not used if `actions/checkout@v2` is used with `fetch-depth: 0`. Is there a way to have the cache work properly when the full git history is pulled?

**Comment by cpuguy83** [comment]
Have you tried adding `.git` to your .dockerignore?

**Comment by bekriebel** [comment]
> Have you tried adding `.git` to your .dockerignore?

I actually need `.git` in my docker workspace, but in a later step where I COPY it in. It isn't copied in any of the earlier steps, so I would expect the cache to be valid up until that COPY step.

**Comment by bekriebel** [comment]
🤦 nevermind, I think I see what is happening. I'm using a multi-stage docker build. Any change that is causing the first stage to be rebuilt is invalidating the cache. It has nothing to do with the fetch-depth, scope, or the `.git` folder specifically.

I had assumed this was caching the whole docker environment including steps for non-final stages - it looks like the only thing that is cached is the final stage image, so anything that requires any step of an earlier stage to be rebuilt appears to cause the cache to not add any savings.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -600,7 +579,7 @@ jobs:
         id: cache
         with:
           path: /tmp/.buildx-cache
-          key: ${{ runner.os }}-buildx-ghcache-${{ github.sha }}
+          key: ${{ runner.os }}-buildx-local-${{ github.sha }}
           restore-keys: |
             ${{ runner.os }}-buildx-ghcache-
```
Comment (pjonsson) [inline .github/workflows/ci.yml:584]: Should the restore-keys have been updated from "ghcache" to "local" as well?

**Pull request merged**